lib/Semantic_check.ml

```diff
@@ -1,2163 +0,0 @@
-        semantic_error ~loc
-          ( "Branches of conditional need to have the same return type. \
-             Instead, found return types "
-          ^ pretty_print_returntype rt1
-          ^ " and "
-          ^ pretty_print_returntype rt2
-          ^ "." )
-      else Complete rt1
-  | Incomplete rt1, Incomplete rt2 | Complete rt1, Incomplete rt2 ->
-      if rt1 <> rt2 then
-        semantic_error ~loc
-          ( "Branches of conditional need to have the same return type. \
-             Instead, found return types "
-          ^ pretty_print_returntype rt1
-          ^ " and "
-          ^ pretty_print_returntype rt2
-          ^ "." )
-      else Incomplete rt1
-  | AnyReturnType, NoReturnType
-   |NoReturnType, AnyReturnType
-   |NoReturnType, NoReturnType ->
-      NoReturnType
-  | AnyReturnType, Incomplete rt
-   |Complete rt, NoReturnType
-   |NoReturnType, Incomplete rt
-   |Incomplete rt, NoReturnType ->
-      Incomplete rt
-  | NoReturnType, Complete rt
-   |Complete rt, AnyReturnType
-   |Incomplete rt, AnyReturnType
-   |AnyReturnType, Complete rt ->
-      Complete rt
-  | AnyReturnType, AnyReturnType -> AnyReturnType
-
-let check_fresh_variable_basic id is_nullary_function =
-  (* For some strange reason, Stan allows user declared identifiers that are
-   not of nullary function types to clash with nullary library functions.
-   No other name clashes are tolerated. Here's the logic to
-   achieve that. *)
-  let _ =
-    if
-      is_stan_math_function_name id.name
-      && ( is_nullary_function
-         || get_stan_math_function_return_type_opt id.name [] = None )
-    then
-      let error_msg =
-        String.concat " " ["Identifier "; id.name; " clashes with primitive."]
-      in
-      semantic_error ~loc:id.id_loc error_msg
-  in
-  match Symbol_table.look vm id.name with
-  | Some _ ->
-      let error_msg =
-        String.concat " " ["Identifier "; id.name; " is already in use."]
-      in
-      semantic_error ~loc:id.id_loc error_msg
-  | None -> ()
-
-let check_fresh_variable id is_nullary_function =
-  List.iter
-    (fun name -> check_fresh_variable_basic name is_nullary_function)
-    (probability_distribution_name_variants id)
-
-(* TODO: the following is very ugly, but we seem to need something like it to
-   reproduce the (strange) behaviour in the current Stan that local variables
-   have a block level that is determined by what has been assigned to them
-   rather than by where they were declared. I'm not sure that behaviour makes
-   sense unless we use static analysis as well to make sure these assignments
-   actually get evaluated in that phase. *)
-let update_originblock name ob =
-  match Symbol_table.look vm name with
-  | Some (old_ob, ut) ->
-      let new_ob = lub_originblock [ob; old_ob] in
-      Symbol_table.unsafe_replace vm name (new_ob, ut)
-  | _ -> fatal_error ()
-
-(* The actual semantic checks for all AST nodes! *)
-let rec semantic_check_program p =
```

Comment:
You can destructure the record above or refer to `p.functionblock`

Reply:
Could you explain what you would like to see changed here? Not sure I understand.